jetstream_pt/cache_manager.py

```diff
@@ -205,23 +499,133 @@ def empty(cls, shape, device, bf16_enable, env):
     v_quant, vscale = self.quantize(xv)
-    if self.env.ring_buffer:
+
+    if self.env.lazy_cache_update:
+      if self.env.new_cache_stacked:
+        self.new_ks[layer_id, ...] = k_quant
+        self.new_vs[layer_id, ...] = v_quant
+        self.new_k_scaler[layer_id, ...] = kscale
+        self.new_v_scaler[layer_id, ...] = vscale
+      else:
+        if self.env.generate_cache_stacked:
+          self.new_ks.append(k_quant)
+          self.new_vs.append(v_quant)
+          self.new_k_scaler.append(kscale)
+          self.new_v_scaler.append(vscale)
+        else:
+          self.new_ks = k_quant
+          self.new_vs = v_quant
+          self.new_k_scaler = kscale
+          self.new_v_scaler = vscale
+    elif self.env.ring_buffer:
       self.cache_k[:, :, self.input_pos, :] = k_quant
       self.cache_v[:, :, self.input_pos, :] = v_quant
       self.k_scaler[:, :, self.input_pos, :] = kscale
       self.v_scaler[:, :, self.input_pos, :] = vscale
     else:
-      batch = jnp.arange(self.env.batch_size)
-      self.cache_k[batch, :, self.input_pos, :] = k_quant.squeeze(2)
-      self.cache_v[batch, :, self.input_pos, :] = v_quant.squeeze(2)
-      self.k_scaler[batch, :, self.input_pos, :] = kscale.squeeze(2)
-      self.v_scaler[batch, :, self.input_pos, :] = vscale.squeeze(2)
-    return self.cache_k, self.cache_v, self.k_scaler, self.v_scaler
+      # We don't handle left aligned but lazy_cache_update=False
+      self.cache_k[self.batch, :, self.input_pos, :] = k_quant.squeeze(2)
+      self.cache_v[self.batch, :, self.input_pos, :] = v_quant.squeeze(2)
+      self.k_scaler[self.batch, :, self.input_pos, :] = kscale.squeeze(2)
+      self.v_scaler[self.batch, :, self.input_pos, :] = vscale.squeeze(2)
+
+    return (
+        self.cache_k,
+        self.cache_v,
+        k_quant,
+        v_quant,
+        self.k_scaler,
+        self.v_scaler,
+        kscale,
+        vscale,
+    )
+
+  def finalize(self):
+    if not self.env.lazy_cache_update:
+      return
+    if self.env.ring_buffer:
+      # Assume no cache stack for ring buffer
+      self.cache_k._elem = self.cache_k._elem.at[..., self.input_pos, :].set(
+          self.new_ks._elem
+      )
+      self.cache_v._elem = self.cache_v._elem.at[..., self.input_pos, :].set(
+          self.new_vs._elem
+      )
+    else:
+      if self.env.generate_cache_stacked:
+        layer, b, head, _, dim = self.cache_k.shape
+        if self.env.new_cache_stacked:
+          # new kv scaler also has to go through shard_map instead of indexing because it needs to reshape to (batch, layer) which mess up with the data
+          caches = [
+              self.cache_k._elem,
+              self.cache_v._elem,
+              self.new_ks._elem,
+              self.new_vs._elem,
+              self.k_scaler._elem,
+              self.v_scaler._elem,
+              self.new_k_scaler,
+              self.new_v_scaler,
+          ]
+          (
+              self.cache_k._elem,
```

Comment:
`._elem` seems spurious; as `_elem` is already a jax array.

So it's either: `x._elem = foo(jax_array_inputs)` OR `x = call_jax(foo, torch_tensor_inputs)`

Reply:
Make sense, decided to remove _elem since it's violating lint anyway.